pkg/detector/detector.go

```diff
@@ -0,0 +1,218 @@
+	defaultNotReadyTickThreshold        = 6
+	defaultPauseBetweenTermination      = time.Minute * 10
+
+	nodeNotReadyDuration = time.Second * 30
+
+	annotationNodeNotReadyTick = "giantswarm.io/node-not-ready-tick"
+	labelNodeRole              = "role"
+	labelNodeRoleMaster        = "master"
+)
+
+type Config struct {
+	Logger    micrologger.Logger
+	K8sClient client.Client
+
+	// MaxNodeTerminationPercentage defines a maximum percentage of nodes that will be returned as 'marked for termination'
+	// ie: if the value is 0.5 and cluster have 10 nodes, than `DetectBadNodes`can only return maximum of 5 nodes
+	// marked for termination at single run
+	MaxNodeTerminationPercentage float64
+	// NotReadyTickThreshold defines a how many times the node must bee seen as NotReady in order to return it as 'marked for termination'
+	NotReadyTickThreshold int
+	// PauseBetweenTermination defines a pause between 2 intervals where node termination can occur.
+	// This is a safeguard to prevent nodes being terminated over and over or to not terminate too much at once.
+	// ie: if the value is 5m it means once it returned nodes for termination it wont return another nodes for another 5 min.
+	PauseBetweenTermination time.Duration
+}
+
+type Detector struct {
+	logger    micrologger.Logger
+	k8sClient client.Client
+
+	maxNodeTerminationPercentage float64
+	notReadyTickThreshold        int
+	pauseBetweenTermination      time.Duration
+}
+
+func NewDetector(config Config) (*Detector, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.MaxNodeTerminationPercentage == 0 {
+		config.MaxNodeTerminationPercentage = defaultMaxNodeTerminationPercentage
+	}
+	if config.NotReadyTickThreshold == 0 {
+		config.NotReadyTickThreshold = defaultNotReadyTickThreshold
+	}
+	if config.PauseBetweenTermination == 0 {
+		config.PauseBetweenTermination = defaultPauseBetweenTermination
+	}
+
+	d := &Detector{
+		logger:    config.Logger,
+		k8sClient: config.K8sClient,
+
+		notReadyTickThreshold:   config.NotReadyTickThreshold,
+		pauseBetweenTermination: config.PauseBetweenTermination,
+	}
+
+	return d, nil
+}
+
+// DetectBadNodes will return list of nodes that should be terminated which in documentation terminology is used as 'marked for termination'.
+func (d *Detector) DetectBadNodes(ctx context.Context) ([]corev1.Node, error) {
+	var nodeList corev1.NodeList
+	{
+		err := d.k8sClient.List(ctx, &nodeList)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	nodesToTerminate, err := d.detectBadNodes(ctx, nodeList.Items)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// remove additional master nodes to avoid multiple master node termination at the same time
+	nodesToTerminate = d.removeMultipleMasterNodes(ctx, nodesToTerminate)
```

Comment:
what if the failed node is the only master for a cluster?
Do we want to terminate it anyway?

Reply:
If the master node is not ready then it kinda makes sense as some important components like `kiam-server` are running on master node and should be running

But we can have a bigger discussion on this